shared/chat/conversation/normal/container.tsx

```diff
@@ -1,54 +1,79 @@
 import * as C from '@/constants'
+import * as T from '@/constants/types'
 import * as React from 'react'
 import Normal from '.'
 import {OrangeLineContext} from '../orange-line-context'
 import {FocusProvider, ScrollProvider} from './context'
 
+const noOrd = T.Chat.numberToOrdinal(-1)
+const caughtUpOrd = T.Chat.numberToOrdinal(0)
 // Orange line logic:
 // While looking at a thread the line should be static
 // If you aren't active (backgrounded on desktop) the orange line will appear above new content
 // If you are active and new items get added the orange line will be consistent, either where it was on first
 // mount or not there at all (active and new items come)
 // Handle mark as unread
 const useOrangeLine = () => {
+  const orangeLineRef = React.useRef(noOrd)
+
   const conversationIDKey = C.useChatContext(s => s.id)
-  const readMsgID = C.useChatContext(s => s.meta.readMsgID)
-  const maxMsgID = C.useChatContext(s => s.meta.maxMsgID)
-  const active = C.useActiveState(s => s.active)
-  const reinitValue = maxMsgID > readMsgID ? readMsgID : 0
-  const orangeLineRef = React.useRef(reinitValue)
   const lastCIDRef = React.useRef(conversationIDKey)
-  const lastReadMsgIDRef = React.useRef(readMsgID)
-  const metaGoodRef = React.useRef(readMsgID > 0)
+  const convoChanged = lastCIDRef.current !== conversationIDKey
+  lastCIDRef.current = conversationIDKey
 
-  // meta not ready yet
-  if (readMsgID < 0) {
-    return 0
-  }
+  const active = C.useActiveState(s => s.active)
+  const lastActiveRf = React.useRef(active)
+  const activeChanged = lastActiveRf.current !== active
+  const wentInactive = !active && activeChanged
+  lastActiveRf.current = active
 
-  // init on first good met
-  if (!metaGoodRef.current) {
-    metaGoodRef.current = true
-    lastReadMsgIDRef.current = readMsgID
-    orangeLineRef.current = reinitValue
-  }
+  const noExisting = orangeLineRef.current === noOrd
 
-  // convo changed? reset
-  if (lastCIDRef.current !== conversationIDKey) {
-    lastCIDRef.current = conversationIDKey
-    lastReadMsgIDRef.current = readMsgID
-    orangeLineRef.current = reinitValue
-  }
+  const readMsgID = C.useChatContext(s => {
+    const {readMsgID} = s.meta
+    return readMsgID
+  })
+  const lastReadMsgIDRef = React.useRef(readMsgID)
+  // mark as unread does this
+  const readMsgWentBackwards = readMsgID > 0 && readMsgID < lastReadMsgIDRef.current
+  lastReadMsgIDRef.current = readMsgID
+
+  // only search for an orange line if we need it
+  const needToGetOrangeLine = convoChanged || noExisting || wentInactive || readMsgWentBackwards
+
+  const storeOrangeLine = C.useChatContext(s => {
+    // don't do a search which could be expensive if we don't need it
+    if (!needToGetOrangeLine) return noOrd
+    const {readMsgID, maxMsgID} = s.meta
+    if (readMsgID <= 0) return noOrd
+    if (maxMsgID > readMsgID) {
+      const mm = s.messageMap
+      // find a good ordinal
```

Comment:
i thought we could skip this but we do need this in some cases so we only do it when we must